### DISCUSSION TITLE
単体テストにおいて、言語の切り替え処理を共通化する

### DISCUSSION THREAD

**Discussion by Takym** [cocoa-mhlw/cocoa#275] [Ideas]
（#269、#272 から）

* 共通化する方法として #272 では `IClassFixture` を使っています。（原案）

**Comment by keiji** [comment]
ぼくは提案手法には反対の立場です。今回のケースでは言語に依存したテストは2件のみ。また、言語の書き換えが必要なスコープはクラス単位でなくてメソッド単位で足るものです。

IClassFixtureを使った場合、クラス単位で言語設定の書き換えをすることになります。本来1つのメソッドしか言語設定を書き換える必要がないのにクラス単位で指定して書き換えてしまうことで、クラス全体が言語に依存するテストであるという誤ったメッセージを伝える可能性があると考えています。

**Reply by Takym** [reply]
では、下記のフィールドを変更する静的関数を作成し、言語に依存したテストから呼び出すのはどうでしょうか？

* `CultureInfo.DefaultThreadCurrentCulture`
* `CultureInfo.DefaultThreadCurrentUICulture`
* `CultureInfo.CurrentCulture`
* `CultureInfo.CurrentUICulture`

`CurrentCulture`/`CurrentUICulture` を両方変更するのは、(1)言語に依存する全ての場所から使用できる様にするため、(2)複数の言語を混在させないためです。

**Reply by keiji** [reply]
これがアプリ側のコードであれば、齟齬が生まれて予期せぬ動作を防ぐために念のためにやっておきましょうか。と、言うと思うのですが、今回のような単体テストではやらない方がいいとぼくは思います。

今回の場合、当該メソッドで`CurrentUICulture`を書き換えることは、「このテストはロケールに依存する」「UI関係のCultureに依存する」テストの前提条件を伝える役割もあります。

`CurrentCulture`を書き換えてもテストの結果には影響しませんし、`DefaultThread...`に至ってはデフォルトで`null`がセットされています。そういう本来、書き換える必要のない設定を共通化の名目で書き換えてしまうことで、テストの前提条件がほやけてしまうことは避けた方がいいとぼくは考えています。

また、必要はない操作を共通化すると後々に弊害があるとも考えています。
最初は「念のため、CurrentCultureやDefaultThread...も書き換えておこう」と言う程度の動機だったものが、時間がたつと「既存のコードでCurrentCultureも書き換えるようになっているから、（理由はわからないけど）書き換えておいた方がいいだろう」に変わります。さらに時間が流れて、共通処理を変更する必要が出てきたときに「どう考えても不要な処理なのに、なぜこうしているのか誰もわからない」となり「やっぱり不要な処理だった」という結論が出るまでに本来必要のない時間を消費する可能性があります。

そのようなわけで、現時点の規模では、単体テストにおいて言語の切り換え処理を共通化しない方がいいとぼくは考えます。

**Comment by Takym** [comment]
## 代案１
[こちら](https://github.com/YigtyORG/TakymLib/blob/master/tests/DependsOn.cs)の様に `DependsOn` クラスを作成し、その中に静的関数 `CurrentUICulture` を作成します。

1. `DependsOn.CurrentUICulture()` と書くことにより、何に依存しているのか分かり易くなります。
    * （コメントを付与せずに済みます。）
2. 言語の切り替え処理を変更する場合は、一箇所のみを書き換えるだけで済みます。

**Reply by keiji** [reply]
今回のケースでは言語に依存したテストは2件だけです。仕組みを作って共通化するメリットに乏しいと考えます。

共通化というのは、同じコードがn件存在して、さらに今後も増えることが予想される場合に検討するのがよいと考えています。閾値nは人それぞれですが、少なくとも現状の２件ではないとぼくは思います。

もちろん、@Takym さんがすでにCOCOAのテストコードを拡充してくださっていて、実際に同様の処理に遭遇しており、言語の切り替え処理を共通化する必要性を感じている。と言うことであれば、ぜひ拡充したテストと合わせてPull Requestをいただければと思います。

**Reply by Takym** [reply]
増えた後に共通化作業を行うのは大変です。増える前であれば少ない量のコードを修正するだけで済みますが、増えた後では多い量のコードを修正した上漏れがないか確認する必要が出てきます。今現在、今後増えないと予想しても増える可能性は存在します。僕の経験では、共通化する必要性が出た時には既に修正すべき箇所が多いです。

**Reply by keiji** [reply]
繰り返しになりますが、COCOAのテストコードを拡充する取り組みを始めていて、実際に同様の処理に遭遇しており、言語の切り替え処理を共通化する必要性を感じている。と言うことであれば、ぜひ拡充したテストと合わせてご提案ください。テストが充実することは大歓迎です。
